**Nama    : Della Shanda Anggrivani**

**NIM     : 24060121120024**

**Dataset : Glass Identification**

**Link    : https://archive.ics.uci.edu/dataset/42/glass+identification**

**1. Melakukan eksplorasi menggunakan algoritma klasifikasi lain pada dataset Iris**

In [53]:
# Import library yang dibutuhkan
import pandas as pd

In [57]:
path = "http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']

dataset = pd.read_csv(path, names=names)

In [58]:
dataset

sepal-length  sepal-width  petal-length  petal-width           class
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]

In [59]:
dataset.shape

(150, 5)

Pada kode diatas dihasilkan 150 jumlah baris data dalam dataset, serta 5 kolom atau fitur dalam dataset, yaitu sepal-length, sepal-width, petal-length, petal-width, dan class

In [60]:
# Import library yang dibutuhkan
from sklearn import model_selection

In [61]:
array = dataset.values

X = array[:,0:4]
Y = array[:,4]

validation_size = 0.20

# untuk mengacak data
seed = 7

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size,random_state=seed)

Code diatas adalah contoh proses persiapan data untuk pembelajaran mesin menggunakan pustaka scikit-learn di Python. Penjelasan code:

1. `array = dataset.values`: Mengonversi dataset ke dalam bentuk array. Jika `dataset` adalah objek DataFrame dari pustaka pandas, maka metode `values` akan mengonversinya ke dalam array NumPy. Dengan menggunakan indeksing, kita dapat memisahkan array menjadi dua bagian: `X` (matriks fitur) dan `Y` (vektor target).

2. `X = array[:, 0:4]`: Memilih kolom 0 hingga 3 (indeks 0, 1, 2, dan 3) dari array sebagai matriks fitur `X`.

3. `Y = array[:, 4]`: Memilih kolom dengan indeks 4 dari array sebagai vektor target `Y`.

4. `validation_size = 0.20`: Menetapkan ukuran data validasi sebesar 20% dari total data. Hal ini akan digunakan dalam pemisahan data latih dan data validasi.

5. `seed = 7`: Menetapkan seed untuk mengacak data. Seed digunakan agar hasil pengacakan dapat direproduksi.

6. `X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)`: menggunakan fungsi `train_test_split` dari modul `model_selection` di scikit-learn untuk memisahkan dataset menjadi data latih (`X_train` dan `Y_train`) dan data validasi (`X_validation` dan `Y_validation`). Parameter `test_size` menentukan ukuran data validasi (dalam hal ini, 20%), dan `random_state` adalah seed untuk pengacakan.


In [62]:
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (120, 4)
X_validation : (30, 4)
Y_train : (120,)
Y_validation : (30,)


Code di atas digunakan untuk mencetak dimensi (shape) dari matriks fitur dan vektor target setelah dilakukan pembagian data menjadi data latih (`X_train` dan `Y_train`) serta data validasi (`X_validation` dan `Y_validation`). Pembahasan code:

1. `print("X_train :", X_train.shape)`: Menampilkan dimensi dari matriks fitur data latih (`X_train`). Hasilnya adalah `(120, 4)`, yang berarti terdapat 120 baris dan 4 kolom dalam matriks fitur `X_train`. Setiap baris mewakili satu sampel data latih, dan setiap kolom mewakili satu fitur.

2. `print("X_validation :", X_validation.shape)`: Menampilkan dimensi dari matriks fitur data validasi (`X_validation`). Hasilnya adalah `(30, 4)`, yang berarti terdapat 30 baris dan 4 kolom dalam matriks fitur `X_validation`. Seperti pada `X_train`, setiap baris mewakili satu sampel data validasi.

3. `print("Y_train :", Y_train.shape)`: Menampilkan dimensi dari vektor target data latih (`Y_train`). Hasilnya adalah `(120,)`, yang berarti terdapat 120 elemen dalam vektor target `Y_train`. Setiap elemen mewakili label atau kelas dari satu sampel data latih.

4. `print("Y_validation :", Y_validation.shape)`: Menampilkan dimensi dari vektor target data validasi (`Y_validation`). Hasilnya adalah `(30,)`, yang berarti terdapat 30 elemen dalam vektor target `Y_validation`. Seperti pada `Y_train`, setiap elemen mewakili label atau kelas dari satu sampel data validasi.


In [63]:
seed = 7

scoring = 'accuracy'

Code di atas menetapkan dua variabel, yaitu seed dan scoring.

1. seed = 7: Mendefinisikan seed yang akan digunakan dalam beberapa fungsi yang melibatkan pengacakan atau pembagian data. Seed digunakan untuk mengontrol hasil acak sehingga hasilnya dapat direproduksi. .

2. scoring = 'accuracy': Mendefinisikan metrik evaluasi yang akan digunakan pada saat validasi model. Dalam konteks ini, metrik evaluasi yang digunakan adalah akurasi (accuracy). Akurasi mengukur sejauh mana model dapat memprediksi dengan benar. Untuk tugas klasifikasi, akurasi adalah rasio antara jumlah prediksi yang benar dengan total jumlah prediksi.

In [64]:
# Import library berisi algoritma klasifikasi yang akan digunakan
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [65]:
# Spot Check Algorithms
models = []

models.append(('Tree', DecisionTreeClassifier()))
models.append(('RandomForest', RandomForestClassifier()))

Code tersebut merupakan bagian dari proses "Spot Checking Algorithms" atau pemeriksaan algoritma untuk memilih model terbaik dari berbagai algoritma klasifikasi.

1. Pembuatan List Models,
baris ini membuat sebuah list kosong yang nantinya akan diisi oleh berbagai model klasifikasi yang akan diuji. Setiap model akan ditambahkan ke dalam list menggunakan metode append()

2. Spot Check Algorithms,
baris ini menambahkan dua model ke dalam list models. Model pertama adalah Decision Tree (pohon keputusan), dan model kedua adalah Random Forest.


In [66]:
# evaluate each model in turn
results = [] #menyimpan hasil evaluasi

names = [] #menyimpan nama model

for name, model in models: #mengambil nama dan model dari daftar models tadi

  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)

  # menghitung nilai akurasi dari setiap iterasi
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)

  results.append(cv_results)

  names.append(name)

  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

Tree: 0.950000 (0.076376)
RandomForest: 0.966667 (0.040825)


Penjelasan code :

1. **Inisialisasi Variabel Hasil dan Nama:**
   ```python
   results = [] # Menyimpan hasil evaluasi
   names = [] # Menyimpan nama model
   ```
   Baris ini membuat dua list kosong (`results` dan `names`) yang akan digunakan untuk menyimpan hasil evaluasi dari setiap model dan nama modelnya.

2. **Looping untuk Setiap Model:**
   ```python
   for name, model in models:
   ```
   Baris ini memulai loop yang akan berjalan untuk setiap model dalam list `models`. Dalam setiap iterasi, `name` akan berisi nama model, dan `model` akan berisi objek model yang akan dievaluasi.

3. **Inisialisasi K-Fold Cross-Validation:**
   ```python
   kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
   ```
   Baris ini membuat objek KFold untuk melakukan validasi silang dengan 10 lipatan. `shuffle=True` menunjukkan bahwa data akan diacak sebelum dibagi menjadi lipatan.

4. **Cross-Validation dan Penyimpanan Hasil:**
   ```python
   cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
   ```
   Baris ini menggunakan `cross_val_score` untuk melakukan validasi silang pada model dengan menggunakan K-Fold yang telah diinisialisasi sebelumnya. Hasilnya disimpan dalam variabel `cv_results`.

5. **Penyimpanan Hasil dan Cetak Informasi:**
   ```python
   results.append(cv_results)
   names.append(name)
   msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
   print(msg)
   ```
   Baris ini menyimpan hasil evaluasi (`cv_results`) ke dalam list `results` dan nama model ke dalam list `names`. Selanjutnya, pesan (`msg`) berisi informasi tentang nama model, rata-rata akurasi, dan deviasi standar akurasi. Pesan ini kemudian dicetak ke layar.



Hasil evaluasi yang dicetak adalah rata-rata akurasi dan deviasi standar dari 10-fold cross-validation untuk setiap model. Berikut adalah penjelasan hasilnya:

1. **Decision Tree (`Tree`):**
   - Rata-rata akurasi: 0.950000
   - Deviasi standar: 0.076376

   Hasil ini menunjukkan bahwa model Decision Tree memiliki rata-rata akurasi sekitar 95% dan deviasi standar sekitar 0.076. Ini berarti model tersebut cukup konsisten dalam melakukan prediksi, dan akurasi rata-ratanya adalah sekitar 95%.

2. **Random Forest (`RandomForest`):**
   - Rata-rata akurasi: 0.966667
   - Deviasi standar: 0.040825

   Hasil ini menunjukkan bahwa model Random Forest memiliki rata-rata akurasi sekitar 96.67% dan deviasi standar sekitar 0.041. Random Forest memiliki akurasi rata-rata yang sedikit lebih tinggi dibandingkan dengan Decision Tree dan juga menunjukkan deviasi standar yang lebih rendah, menunjukkan konsistensi yang baik dalam prediksi.

Secara umum, hasil ini menunjukkan bahwa Random Forest memberikan kinerja yang sedikit lebih baik daripada Decision Tree untuk dataset dan tugas klasifikasi yang sedang diuji.

In [67]:
# Make predictions on validation dataset
RF = RandomForestClassifier()

# melatih svm menggunakan data latih
RF.fit(X_train, Y_train)

# menguji model klasifikasi yang dihasilkan dari proses pelatihan
# data dengan testing data
predictions = RF.predict(X_validation)

print("Accuracy:", accuracy_score(Y_validation, predictions))
print("\nConfusion Matrix:\n", confusion_matrix(Y_validation, predictions))
print("\nClassification Report:\n", classification_report(Y_validation, predictions))

Accuracy: 0.8666666666666667

Confusion Matrix:
 [[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]

Classification Report:
                  precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



Code diatas digunakan untuk membuat prediksi pada dataset validasi menggunakan model klasifikasi Random Forest yang telah dilatih sebelumnya. Penjelasan code :

1. **Inisialisasi dan Pelatihan Model:**
   ```python
   RF = RandomForestClassifier()
   RF.fit(X_train, Y_train)
   ```
   Baris pertama membuat objek dari kelas `RandomForestClassifier` dan menetapkannya ke variabel `RF`. Baris kedua melatih model tersebut menggunakan data latih (`X_train` dan `Y_train`).

2. **Membuat Prediksi pada Dataset Validasi:**
   ```python
   predictions = RF.predict(X_validation)
   ```
   Baris ini menggunakan model yang telah dilatih untuk membuat prediksi pada dataset validasi (`X_validation`). Hasil prediksi disimpan dalam variabel `predictions`.

3. **Evaluasi Performa Model pada Dataset Validasi:**
   ```python
   print("Accuracy:", accuracy_score(Y_validation, predictions))
   print("\nConfusion Matrix:\n", confusion_matrix(Y_validation, predictions))
   print("\nClassification Report:\n", classification_report(Y_validation, predictions))
   ```
   Baris ini mencetak beberapa metrik evaluasi performa model pada dataset validasi. Metrik tersebut meliputi akurasi (`accuracy`), `confusion matrix`, dan laporan klasifikasi (`classification report`).

   - `accuracy_score`: Menghitung akurasi prediksi model.
   - `confusion_matrix`: Menampilkan matriks, yang memberikan informasi tentang jumlah prediksi yang benar dan salah untuk setiap kelas.
   - `classification_report`: Menampilkan laporan klasifikasi yang mencakup precision, recall, dan f1-score untuk setiap kelas, serta rata-rata akurasi dan f1-score secara keseluruhan.



Hasil evaluasi performa model pada dataset validasi mencakup tiga komponen utama: akurasi (`accuracy`),`confusion matrix`, dan laporan klasifikasi (`classification report`). Penjelasan hasil :

1. **Akurasi (Accuracy):**
   ```
   Accuracy: 0.8666666666666667
   ```
   Akurasi adalah metrik yang mengukur sejauh mana model dapat memprediksi dengan benar. Dalam konteks ini, akurasi model pada dataset validasi adalah sekitar 86.67%, yang berarti sekitar 86.67% dari prediksi model benar.

2. **Confusion Matrix:**

```
Confusion Matrix:
[[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]
```

- **Baris pertama:** Mewakili kelas Iris-setosa.
- **Baris kedua:** Mewakili kelas Iris-versicolor.
- **Baris ketiga:** Mewakili kelas Iris-virginica.

- **Kolom pertama:** Mewakili prediksi untuk kelas Iris-setosa.
- **Kolom kedua:** Mewakili prediksi untuk kelas Iris-versicolor.
- **Kolom ketiga:** Mewakili prediksi untuk kelas Iris-virginica.

Dengan menggunakan confusion matrix, kita dapat melihat:

1. **Kelas Iris-setosa (Baris Pertama):**
   - Terdapat 7 prediksi yang benar (True Positive).
   - Tidak ada prediksi yang salah untuk kelas Iris-setosa (False Positive dan False Negative adalah 0).

2. **Kelas Iris-versicolor (Baris Kedua):**
   - Terdapat 10 prediksi yang benar (True Positive).
   - Terdapat 2 prediksi yang salah, yaitu 2 data Iris-versicolor yang salah diprediksi sebagai Iris-virginica (False Negative).

3. **Kelas Iris-virginica (Baris Ketiga):**
   - Terdapat 9 prediksi yang benar (True Positive).
   - Terdapat 2 prediksi yang salah, yaitu 2 data Iris-virginica yang salah diprediksi sebagai Iris-versicolor (False Negative).


3. **Laporan Klasifikasi (Classification Report):**
   ```
                  precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30
   ```
   Laporan klasifikasi memberikan informasi lebih rinci tentang precision, recall, dan f1-score untuk setiap kelas, serta rata-rata akurasi dan f1-score secara keseluruhan. Precision mengukur seberapa baik model dapat mengidentifikasi instance positif, recall mengukur sejauh mana model dapat mendeteksi semua instance positif yang seharusnya, dan f1-score adalah harmonic mean dari precision dan recall.

   - `Iris-setosa`: Precision, recall, dan f1-score sebesar 1.00, menunjukkan performa sempurna untuk kelas ini.
   - `Iris-versicolor` dan `Iris-virginica`: Precision, recall, dan f1-score sekitar 0.83 dan 0.82, menunjukkan performa yang baik, meskipun tidak sempurna.

   Bagian akhir dari laporan klasifikasi (`accuracy`, `macro avg`, dan `weighted avg`) memberikan rata-rata akurasi dan f1-score, baik secara rata-rata sederhana (macro avg) maupun berbobot (weighted avg). Dalam kasus ini, akurasi rata-rata adalah sekitar 87%.

**2. Melakukan evaluasi algoritma dengan dataset Glass Identification, menggunakan algoritma KNN, NB, dan SVM**

Import libraries

In [69]:
# Import library yang dibutuhkan
import pandas as pd

Membaca dan mempersiapkan dataset

In [70]:
# Variabel names berisi daftar nama kolom/feature yang akan digunakan untuk DataFrame yang akan dibaca.
names = ['RI', 'Na', 'Mg', 'AI', 'Si', 'K', 'Ca', 'Ba', 'Fe', 'Type of glass']

# Dataset yang digunakan
dataset = pd.read_csv('glass.data', names=names)

Melihat isi dataset

In [71]:
dataset

RI     Na    Mg    AI     Si     K    Ca    Ba   Fe  Type of glass
1    1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.00  0.0              1
2    1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.00  0.0              1
3    1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.00  0.0              1
4    1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.00  0.0              1
5    1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.00  0.0              1
..       ...    ...   ...   ...    ...   ...   ...   ...  ...            ...
210  1.51623  14.14  0.00  2.88  72.61  0.08  9.18  1.06  0.0              7
211  1.51685  14.92  0.00  1.99  73.06  0.00  8.40  1.59  0.0              7
212  1.52065  14.36  0.00  2.02  73.42  0.00  8.44  1.64  0.0              7
213  1.51651  14.38  0.00  1.94  73.61  0.00  8.48  1.57  0.0              7
214  1.51711  14.23  0.00  2.08  73.36  0.00  8.62  1.67  0.0              7

[214 rows x 10 columns]

In [72]:
dataset.shape

(214, 10)

Pada kode diatas dihasilkan 214 jumlah baris data dalam dataset, serta 10 kolom atau fitur dalam dataset, yaitu RI, Na, Mg, AI, Si, K, Ca, Ba, Fe, dan Type of glass.

Train test split

In [73]:
# Import library yang dibutuhkan
from sklearn import model_selection

In [74]:
array = dataset.values

X = array[:,0:9]
Y = array[:,9]
Y = Y.astype('int')

validation_size = 0.20

# untuk mengacak data
seed = 7

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size,random_state=seed)

Code diatas adalah contoh proses persiapan data untuk pembelajaran mesin menggunakan pustaka scikit-learn di Python. Penjelasan code:

- array = dataset.values: Mengonversi dataset ke dalam bentuk array. Jika dataset adalah objek DataFrame dari pustaka pandas, maka metode values akan mengonversinya ke dalam array NumPy. Dengan menggunakan indeksing, kita dapat memisahkan array menjadi dua bagian: X (matriks fitur) dan Y (vektor target).

- X = array[:, 0:9]: Memilih kolom 0 hingga 8 dari array sebagai matriks fitur X.

- Y = array[:, 9]: Memilih kolom dengan indeks 9 dari array sebagai vektor target Y.

- Y = Y.astype('int') digunakan untuk mengonversi tipe data vektor target Y menjadi tipe data integer.

- validation_size = 0.20: Menetapkan ukuran data validasi sebesar 20% dari total data. Hal ini akan digunakan dalam pemisahan data latih dan data validasi.

- seed = 7: Menetapkan seed untuk mengacak data. Seed digunakan agar hasil pengacakan dapat direproduksi.

- X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed): menggunakan fungsi train_test_split dari modul model_selection di scikit-learn untuk memisahkan dataset menjadi data latih (X_train dan Y_train) dan data validasi (X_validation dan Y_validation). Parameter test_size menentukan ukuran data validasi (dalam hal ini, 20%), dan random_state adalah seed untuk pengacakan.

In [75]:
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (171, 9)
X_validation : (43, 9)
Y_train : (171,)
Y_validation : (43,)


Code di atas digunakan untuk mencetak dimensi (shape) dari matriks fitur dan vektor target setelah dilakukan pembagian data menjadi data latih (X_train dan Y_train) serta data validasi (X_validation dan Y_validation). Pembahasan code:

print("X_train :", X_train.shape): Menampilkan dimensi dari matriks fitur data latih (X_train). Hasilnya adalah (171, 9), yang berarti terdapat 171 baris dan 9 kolom dalam matriks fitur X_train. Setiap baris mewakili satu sampel data latih, dan setiap kolom mewakili satu fitur.

print("X_validation :", X_validation.shape): Menampilkan dimensi dari matriks fitur data validasi (X_validation). Hasilnya adalah (43, 9), yang berarti terdapat 43 baris dan 9 kolom dalam matriks fitur X_validation. Seperti pada X_train, setiap baris mewakili satu sampel data validasi.

print("Y_train :", Y_train.shape): Menampilkan dimensi dari vektor target data latih (Y_train). Hasilnya adalah (171,), yang berarti terdapat 171 elemen dalam vektor target Y_train. Setiap elemen mewakili label atau kelas dari satu sampel data latih.

print("Y_validation :", Y_validation.shape): Menampilkan dimensi dari vektor target data validasi (Y_validation). Hasilnya adalah (43,), yang berarti terdapat 43 elemen dalam vektor target Y_validation. Seperti pada Y_train, setiap elemen mewakili label atau kelas dari satu sampel data validasi.

In [76]:
seed = 7

scoring = 'accuracy'

Code di atas menetapkan dua variabel, yaitu seed dan scoring.

seed = 7: Mendefinisikan seed yang akan digunakan dalam beberapa fungsi yang melibatkan pengacakan atau pembagian data. Seed digunakan untuk mengontrol hasil acak sehingga hasilnya dapat direproduksi. .

scoring = 'accuracy': Mendefinisikan metrik evaluasi yang akan digunakan pada saat validasi model. Dalam konteks ini, metrik evaluasi yang digunakan adalah akurasi (accuracy). Akurasi mengukur sejauh mana model dapat memprediksi dengan benar. Untuk tugas klasifikasi, akurasi adalah rasio antara jumlah prediksi yang benar dengan total jumlah prediksi.

In [77]:
# Import library berisi algoritma klasifikasi yang akan digunakan
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

In [78]:
# Spot Check Algorithms
models = []

models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))

Code tersebut merupakan bagian dari proses "Spot Checking Algorithms" atau pemeriksaan algoritma untuk memilih model terbaik dari berbagai algoritma klasifikasi.

1. Pembuatan List Models, baris ini membuat sebuah list kosong yang nantinya akan diisi oleh berbagai model klasifikasi yang akan diuji. Setiap model akan ditambahkan ke dalam list menggunakan metode append()

2. Spot Check Algorithms, baris ini menambahkan dua model ke dalam list models. Model pertama adalah KNN, dan model kedua adalah Naive Bayes, dan model ketiga adalah SVM.

In [79]:
# evaluate each model in turn
results = [] #menyimpan hasil evaluasi

names = [] #menyimpan nama model

for name, model in models: #mengambil nama dan model dari daftar models tadi

  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)

  # menghitung nilai akurasi dari setiap iterasi
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring, error_score='raise')

  results.append(cv_results)

  names.append(name)

  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

KNN: 0.671569 (0.125187)
NB: 0.391503 (0.062036)
SVM: 0.350654 (0.077501)


Code tersebut digunakan untuk mengevaluasi performa beberapa model klasifikasi menggunakan validasi silang (cross-validation) dengan 10 lipatan (10-fold cross-validation). Penjelasan :

1. **Inisialisasi Variabel Hasil dan Nama:**
   ```python
   results = [] # Menyimpan hasil evaluasi
   names = [] # Menyimpan nama model
   ```
   Baris ini membuat dua list kosong (`results` dan `names`) yang akan digunakan untuk menyimpan hasil evaluasi dari setiap model dan nama modelnya.

2. **Looping untuk Setiap Model:**
   ```python
   for name, model in models:
   ```
   Baris ini memulai loop yang akan berjalan untuk setiap model dalam list `models`. Dalam setiap iterasi, `name` akan berisi nama model, dan `model` akan berisi objek model yang akan dievaluasi.

3. **Inisialisasi K-Fold Cross-Validation:**
   ```python
   kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
   ```
   Baris ini membuat objek KFold untuk melakukan validasi silang dengan 10 lipatan. `shuffle=True` menunjukkan bahwa data akan diacak sebelum dibagi menjadi lipatan.

4. **Cross-Validation dan Penyimpanan Hasil:**
   ```python
   cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring, error_score='raise')
   ```
   Baris ini menggunakan `cross_val_score` untuk melakukan validasi silang pada model dengan menggunakan K-Fold yang telah diinisialisasi sebelumnya. Hasilnya disimpan dalam variabel `cv_results`.

   - `scoring`: Parameter ini menentukan metrik evaluasi yang akan digunakan. Dalam konteks ini, metriknya adalah akurasi.
   - `error_score='raise'`: Parameter ini menentukan apakah kesalahan dalam validasi silang harus diangkat (`'raise'`) atau diabaikan (`'raise'` akan memunculkan kesalahan jika ada).

5. **Penyimpanan Hasil dan Cetak Informasi:**
   ```python
   results.append(cv_results)
   names.append(name)
   msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
   print(msg)
   ```
   Baris ini menyimpan hasil evaluasi (`cv_results`) ke dalam list `results` dan nama model ke dalam list `names`. Selanjutnya, pesan (`msg`) berisi informasi tentang nama model, rata-rata akurasi, dan deviasi standar akurasi. Pesan ini kemudian dicetak ke layar.



Hasil cetakan tersebut memberikan informasi tentang performa beberapa model klasifikasi yang dievaluasi menggunakan validasi silang dengan 10 lipatan.

1. **Model KNN (K-Nearest Neighbors):**
   ```
   KNN: 0.671569 (0.125187)
   ```
   - Rata-rata akurasi KNN selama 10 lipatan adalah sekitar 67.16%.
   - Deviasi standar sekitar 12.52% menunjukkan variasi akurasi antar lipatan. Semakin kecil deviasi standar, semakin stabil performa model.

2. **Model Naive Bayes (NB):**
   ```
   NB: 0.391503 (0.062036)
   ```
   - Rata-rata akurasi Naive Bayes selama 10 lipatan adalah sekitar 39.15%.
   - Deviasi standar sekitar 6.20% menunjukkan variasi akurasi antar lipatan. Perlu diperhatikan bahwa akurasi yang rendah dan deviasi standar yang kecil menunjukkan bahwa model Naive Bayes mungkin tidak cocok untuk dataset ini atau perlu penanganan lebih lanjut.

3. **Model Support Vector Machine (SVM):**
   ```
   SVM: 0.350654 (0.077501)
   ```
   - Rata-rata akurasi SVM selama 10 lipatan adalah sekitar 35.07%.
   - Deviasi standar sekitar 7.75% menunjukkan variasi akurasi antar lipatan. Seperti halnya dengan Naive Bayes, akurasi yang rendah dan deviasi standar yang kecil menunjukkan bahwa model SVM mungkin tidak cocok untuk dataset ini atau perlu penanganan lebih lanjut.





In [81]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [87]:
from sklearn.preprocessing import StandardScaler

# Inisialisasi model k-NN
knn_model = KNeighborsClassifier(n_neighbors=3)  # n_neighbors adalah jumlah tetangga yang akan digunakan

# Melakukan normalisasi data dengan StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)  # X adalah matriks fitur

# Melatih model k-NN menggunakan data latih
knn_model.fit(X_train, Y_train)

# Melakukan prediksi pada dataset validasi
predictions_knn = knn_model.predict(X_validation)

# Evaluasi performa model k-NN
print("Accuracy:", accuracy_score(Y_validation, predictions_knn))
print("\nConfusion Matrix:\n", confusion_matrix(Y_validation, predictions_knn))
print("\nClassification Report:\n", classification_report(Y_validation, predictions_knn, zero_division=1))


Accuracy: 0.6511627906976745

Confusion Matrix:
 [[ 8  2  0  0  0  0]
 [ 6 14  2  0  0  0]
 [ 1  2  0  0  0  0]
 [ 1  0  0  1  0  0]
 [ 0  0  0  0  0  1]
 [ 0  0  0  0  0  5]]

Classification Report:
               precision    recall  f1-score   support

           1       0.50      0.80      0.62        10
           2       0.78      0.64      0.70        22
           3       0.00      0.00      0.00         3
           5       1.00      0.50      0.67         2
           6       1.00      0.00      0.00         1
           7       0.83      1.00      0.91         5

    accuracy                           0.65        43
   macro avg       0.69      0.49      0.48        43
weighted avg       0.68      0.65      0.64        43



Code tersebut merupakan implementasi dari algoritma k-Nearest Neighbors (k-NN) menggunakan pustaka scikit-learn dalam bahasa pemrograman Python.

1. **Import Library:**
   ```python
   from sklearn.preprocessing import StandardScaler
   ```
   Baris ini mengimpor modul `StandardScaler` dari scikit-learn, yang digunakan untuk normalisasi data.

2. **Inisialisasi Model k-NN:**
   ```python
   knn_model = KNeighborsClassifier(n_neighbors=3)
   ```
   Baris ini membuat objek model k-NN dengan menggunakan `KNeighborsClassifier` dan mengatur parameter `n_neighbors` ke 3. Parameter ini menentukan jumlah tetangga terdekat yang akan dipertimbangkan oleh model saat melakukan prediksi.

3. **Normalisasi Data:**
   ```python
   scaler = StandardScaler()
   X_scaled = scaler.fit_transform(X)
   ```
   Baris ini membuat objek `StandardScaler` dan menggunakan metode `fit_transform` untuk melakukan normalisasi data fitur (`X`). Normalisasi dilakukan agar semua fitur memiliki skala yang serupa dan tidak mendominasi satu sama lain.

4. **Pelatihan Model k-NN Menggunakan Data Latih:**
   ```python
   knn_model.fit(X_train, Y_train)
   ```
   Baris ini melatih model k-NN menggunakan data latih (`X_train`, `Y_train`). Model mempelajari pola dari data latih untuk kemudian digunakan dalam prediksi.

5. **Prediksi pada Dataset Validasi:**
   ```python
   predictions_knn = knn_model.predict(X_validation)
   ```
   Baris ini membuat prediksi menggunakan model k-NN pada dataset validasi (`X_validation`). Model menggunakan informasi yang telah dipelajari selama pelatihan untuk memprediksi kelas dari data validasi.

6. **Evaluasi Performa Model:**
   ```python
   print("Accuracy:", accuracy_score(Y_validation, predictions_knn))
   print("\nConfusion Matrix:\n", confusion_matrix(Y_validation, predictions_knn))
   print("\nClassification Report:\n", classification_report(Y_validation, predictions_knn, zero_division=1))
   ```
   Baris ini mencetak metrik evaluasi performa model k-NN pada dataset validasi:
   - `accuracy_score`: Akurasi prediksi, mengukur sejauh mana prediksi model cocok dengan label aktual.
   - `confusion_matrix`: menunjukkan jumlah prediksi yang benar dan salah untuk setiap kelas.
   - `classification_report`: Laporan klasifikasi yang mencakup precision, recall, dan f1-score untuk setiap kelas, serta rata-rata akurasi dan f1-score secara keseluruhan. `zero_division=1` digunakan untuk menghindari peringatan (warning) jika terdapat pembagian dengan nol dalam perhitungan.


Evaluasi performa model k-NN pada dataset validasi memberikan informasi tentang seberapa baik model tersebut dapat melakukan prediksi pada data yang tidak pernah dilihat selama pelatihan.

1. **Accuracy (Akurasi):**
   ```
   Accuracy: 0.6511627906976745
   ```
   Akurasi mengukur seberapa besar proporsi prediksi yang benar dari total prediksi. Dalam konteks ini, akurasi model k-NN sekitar 65.12%, yang berarti sekitar 65.12% dari semua prediksi yang dibuat oleh model benar.

2. **Confusion Matrix :**
```
[[ 8  2  0  0  0  0]
 [ 6 14  2  0  0  0]
 [ 1  2  0  0  0  0]
 [ 1  0  0  1  0  0]
 [ 0  0  0  0  0  1]
 [ 0  0  0  0  0  5]]
```
- Baris matriks ini mewakili kelas sebenarnya.
- Kolom matriks ini mewakili kelas yang diprediksi oleh model.
- Jumlah diagonal utama menunjukkan jumlah prediksi yang benar untuk setiap kelas.
  
**Analisis Confusion Matrix:**
- Untuk kelas 1, ada 8 prediksi yang benar, 2 prediksi yang salah terhadap kelas 2.
- Untuk kelas 2, ada 14 prediksi yang benar, 6 prediksi yang salah terhadap kelas 1, 2 prediksi yang salah terhadap kelas 3.
- Untuk kelas 3, tidak ada prediksi yang benar, dan 3 prediksi yang salah terhadap kelas 1 dan 2.
- Untuk kelas 5, ada 1 prediksi yang benar, dan 1 prediksi yang salah terhadap kelas 1.
- Untuk kelas 6, ada 1 prediksi yang benar.
- Untuk kelas 7, ada 5 prediksi yang benar.

**Classification Report (Laporan Klasifikasi):**
```
             precision    recall  f1-score   support

           1       0.50      0.80      0.62        10
           2       0.78      0.64      0.70        22
           3       0.00      0.00      0.00         3
           5       1.00      0.50      0.67         2
           6       1.00      0.00      0.00         1
           7       0.83      1.00      0.91         5

accuracy                           0.65        43
macro avg       0.69      0.49      0.48        43
weighted avg    0.68      0.65      0.64        43
```

3. **Analisis Classification Report:**
- **Precision (Presisi):**
  - Presisi untuk kelas 1: 0.50. Dari semua prediksi positif untuk kelas 1, 50% di antaranya adalah benar.
  - Presisi untuk kelas 2: 0.78. Dari semua prediksi positif untuk kelas 2, 78% di antaranya adalah benar.
  - Presisi untuk kelas 3: 0.00. Tidak ada prediksi positif untuk kelas 3.
  - Presisi untuk kelas 5: 1.00. Dari semua prediksi positif untuk kelas 5, 100% di antaranya adalah benar.
  - Presisi untuk kelas 6: 1.00. Dari semua prediksi positif untuk kelas 6, 100% di antaranya adalah benar.
  - Presisi untuk kelas 7: 0.83. Dari semua prediksi positif untuk kelas 7, 83% di antaranya adalah benar.

- **Recall (Sensitivitas):**
  - Recall untuk kelas 1: 0.80. Model berhasil mengidentifikasi 80% dari semua instance kelas 1.
  - Recall untuk kelas 2: 0.64. Model berhasil mengidentifikasi 64% dari semua instance kelas 2.
  - Recall untuk kelas 3: 0.00. Model tidak dapat mengidentifikasi instance kelas 3.
  - Recall untuk kelas 5: 0.50. Model berhasil mengidentifikasi 50% dari semua instance kelas 5.
  - Recall untuk kelas 6: 0.00. Model tidak dapat mengidentifikasi instance kelas 6.
  - Recall untuk kelas 7: 1.00. Model berhasil mengidentifikasi semua instance kelas 7.

- **F1-Score:**
  - F1-score memberikan rata-rata harmonik dari presisi dan recall.
  - F1-score untuk kelas 1: 0.62.
  - F1-score untuk kelas 2: 0.70.
  - F1-score untuk kelas 3: 0.00.
  - F1-score untuk kelas 5: 0.67.
  - F1-score untuk kelas 6: 0.00.
  - F1-score untuk kelas 7: 0.91.

- **Support:**
  - Jumlah instance di setiap kelas.

- **Accuracy (Akurasi):**
  - Akurasi keseluruhan dari model terhadap dataset validasi adalah 65%.
  
- **Macro Avg dan Weighted Avg:**
  - Macro Avg adalah rata-rata dari setiap metrik tanpa mempertimbangkan bobot kelas. Weighted Avg memberikan bobot lebih besar pada kelas yang memiliki lebih banyak instance.

**Kesimpulan:**
- Meskipun akurasi keseluruhan sekitar 65%, performa model bervariasi tergantung pada kelasnya.
- Kelas 1 memiliki presisi yang rendah (50%) namun recall yang tinggi (80%), menunjukkan bahwa model cenderung untuk lebih banyak mengidentifikasi instance kelas 1, meskipun beberapa prediksinya salah.
- Kelas 3 memiliki precision dan recall yang rendah (0.00), menunjukkan bahwa model tidak dapat mengidentifikasi instance kelas 3.
- F1-score dapat memberikan gambaran keseluruhan tentang keseimbangan antara precision dan recall untuk setiap kelas. Sebuah f1-score yang tinggi mengindikasikan keseimbangan yang baik.